### PR TITLE
Allocate space for locals together with stack

### PR DESCRIPTION
**Pull request by chfast**
This move memory management of args and locals to OperandStack where
space for all can be allocated in single go.

### CODE REVIEW

**Comment by axic** [comment]
This seems to be very similar to #358.

What is the story with this?

**Comment by chfast** [comment]
I resurrected this one because I will need it in near future.
Benchmarks are not great, but reviews are welcome anyway.

```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     +0.0298         +0.0302            75            77            75            77
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    +0.0305         +0.0307          1140          1175          1140          1175
fizzy/execute/ecpairing/onepoint_mean                             -0.0227         -0.0225        389763        380929        389707        380931
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   -0.0390         -0.0389            98            94            98            94
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  -0.0325         -0.0325          1442          1396          1442          1396
fizzy/execute/memset/256_bytes_mean                               +0.0142         +0.0143             6             6             6             6
fizzy/execute/memset/60000_bytes_mean                             +0.0272         +0.0272          1385          1422          1385          1422
fizzy/execute/mul256_opt0/input0_mean                             +0.0244         +0.0245            24            25            24            25
fizzy/execute/mul256_opt0/input1_mean                             +0.0228         +0.0227            24            25            24            25
fizzy/execute/ramanujan_pi/33_runs_mean                           -0.0770         -0.0770           125           116           125           116
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0116         -0.0117            85            84            85            84
fizzy/execute/sha1/512_bytes_rounds_16_mean                       -0.0262         -0.0262          1197          1166          1197          1166
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0352         -0.0352            88            85            88            85
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0365         -0.0365          1212          1168          1212          1168
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0312         -0.0312         41338         40047         41338         40046
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.2578         -0.2578             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0742         +0.0742             4             5             4             5
fizzy/execute/micro/factorial/10_mean                             -0.4361         -0.4361             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.4605         -0.4605             1             0             1             0
fizzy/execute/micro/fibonacci/24_mean                             -0.3686         -0.3686          7465          4714          7465          4713
fizzy/execute/micro/host_adler32/1_mean                           -0.2608         -0.2608             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0155         -0.0155             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0074         -0.0074            29            29            29            29
fizzy/execute/micro/spinner/1_mean                                -0.4394         -0.4394             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0629         -0.0629             9             8             9             8
```

**Comment by chfast** [comment]
LTO builds looks a bit better:
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0068         -0.0068            86            85            86            85
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0097         -0.0097          1299          1286          1299          1286
fizzy/execute/ecpairing/onepoint_mean                             -0.0480         -0.0480        434523        413657        434527        413656
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   -0.0511         -0.0511           107           101           107           101
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  -0.0750         -0.0750          1553          1437          1553          1437
fizzy/execute/memset/256_bytes_mean                               -0.0350         -0.0350             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0294         -0.0294          1593          1546          1593          1546
fizzy/execute/mul256_opt0/input0_mean                             -0.0324         -0.0324            28            27            28            27
fizzy/execute/mul256_opt0/input1_mean                             -0.0339         -0.0340            28            27            28            27
fizzy/execute/ramanujan_pi/33_runs_mean                           +0.0383         +0.0383           133           138           133           138
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0422         -0.0422            94            90            94            90
fizzy/execute/sha1/512_bytes_rounds_16_mean                       -0.0388         -0.0388          1314          1263          1314          1263
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0266         -0.0266            96            94            96            94
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0170         -0.0171          1323          1301          1323          1301
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0616         -0.0616         42790         40155         42790         40155
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.3255         -0.3255             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0118         +0.0118             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             -0.4079         -0.4079             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.4315         -0.4315             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             -0.3312         -0.3313          7620          5096          7620          5096
fizzy/execute/micro/host_adler32/1_mean                           -0.2273         -0.2273             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0301         +0.0301             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0828         -0.0828            31            29            31            29
fizzy/execute/micro/spinner/1_mean                                -0.4048         -0.4048             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0260         -0.0260            11            10            11            10
```

**Review by gumb0 — COMMENTED** [review]
Looks good overall, would be nice to add some tests with non-empty locals and args.

**Comment by gumb0** [comment]
Does it make sense to try to increase `small_storage_size` because it now includes not only stack, so might go above more often?

File: lib/fizzy/stack.hpp [hunk 1]
```diff
@@ -69,29 +74,43 @@ class OperandStack
 
     Value* bottom() noexcept { return m_large_storage ? m_large_storage.get() : m_small_storage; }
 
-    const Value* bottom() const noexcept
```
[the remark this answers is not in the document]
Comment (chfast) [reply lib/fizzy/stack.hpp:72]: It's a bug. Fixed in #529, but I want to add also regression tests.

The bug is that `bottom()` returns wrong results (ignores locals) and is used in `rbegin()`.

**Comment by chfast** [comment]
> Does it make sense to try to increase `small_storage_size` because it now includes not only stack, so might go above more often?

I tried with 1K and 2K sizes. The performance is not much better, but it causes stack overflow in some infinite call tests.
I think it is fine to leave it as is because #529 replaces "small storage" with "external storage" anyway. That main point of this PR is to separate storage management for args. locals and stack so we can experiment with storage allocation more easily.

File: lib/fizzy/stack.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (chfast) [reply lib/fizzy/stack.hpp:72]: This is fixed now. It turned out the `bottom()` was only used in `shrink()` which was unused. I removed both.

**Comment by chfast** [comment]
We may consider renaming `OperandStack` type as it now manages:
- operand stack,
- locals,
- arguments,
- "stack space" allocation.

In future it will probably be converted to RAII type and cooperate with "thread execution context" e.g. to bump call depth.


Proposals:
- `ExecutionContext`

**Comment by chfast** [comment]
> > Does it make sense to try to increase `small_storage_size` because it now includes not only stack, so might go above more often?
> 
> I tried with 1K and 2K sizes. The performance is not much better, but it causes stack overflow in some infinite call tests.
> I think it is fine to leave it as is because #529 replaces "small storage" with "external storage" anyway. That main point of this PR is to separate storage management for args. locals and stack so we can experiment with storage allocation more easily.

I also tried the small storage of 256 bytes, but I see no performance difference. So leaving it at 128.

File: lib/fizzy/stack.hpp [hunk 2]
```diff
@@ -61,37 +62,53 @@ class OperandStack
...
+        m_top = m_bottom - 1;
+
+        std::copy(std::begin(args), std::end(args), m_locals);
+        std::fill_n(m_locals + num_args, num_locals, 0);
     }
 
     OperandStack(const OperandStack&) = delete;
     OperandStack& operator=(const OperandStack&) = delete;
 
+    Value& local(size_t index) noexcept { return m_locals[index]; }
```
Comment (gumb0) [inline lib/fizzy/stack.hpp:108]: maybe `assert(m_locals + index < m_bottom)`

File: test/unittests/stack_test.cpp [hunk 3]
```diff
@@ -187,26 +228,39 @@ TEST(operand_stack, large)
...
-    EXPECT_EQ(stack[new_height - 1].i64, 0);
+    for (int expected = max_height - 1; expected >= 0; --expected)
+        EXPECT_EQ(stack.pop().i64, expected);
+    EXPECT_EQ(stack.size(), 0);
+
+    EXPECT_EQ(stack.local(0).i64, 0xa1);
+    EXPECT_EQ(stack.local(1).i64, 0xa2);
+
+    for (unsigned i = 0; i < num_locals; ++i)
+        EXPECT_EQ(stack.local(num_args + i).i64, 0);
```
Comment (gumb0) [inline test/unittests/stack_test.cpp:252]:
```suggestion
    for (unsigned i = num_args; i < num_locals; ++i)
        EXPECT_EQ(stack.local(i).i64, 0);
```

**Review by gumb0 — APPROVED** [review]
Looks good, it would be also nice to test `rbegin() / rend()` for case with locals.

File: test/unittests/stack_test.cpp [hunk 4]
```diff
@@ -171,10 +181,61 @@ TEST(operand_stack, small)
...
+    EXPECT_EQ(stack.local(2).i64, 0xc2);
+    EXPECT_EQ(stack.local(3).i64, 0xc3);
+    EXPECT_EQ(stack.local(4).i64, 0xc4);
+}
+
 TEST(operand_stack, large)
 {
     constexpr auto max_height = 33;
     OperandStack stack({}, 0, max_height);
+    ASSERT_GT(address_diff(&stack, stack.rbegin()), 100) << "not allocated on the heap";
```
Comment (axic) [inline test/unittests/stack_test.cpp:220]: Why is this GT while the others LT?

Comment (chfast) [reply test/unittests/stack_test.cpp:220]: The `100` is arbitrary.

File: lib/fizzy/stack.hpp [hunk 5]
```diff
@@ -61,37 +62,57 @@ class OperandStack
...
-    }
-
 public:
     /// Default constructor.
     ///
     /// Based on @p max_stack_height decides if to use small pre-allocated storage or allocate
     /// large storage.
     /// Sets the top item pointer to below the stack bottom.
-    explicit OperandStack(size_t max_stack_height)
+    OperandStack(span<const Value> args, size_t num_locals, size_t max_stack_height)
```
Comment (axic) [inline lib/fizzy/stack.hpp:81]: Should document here how the stack + args are laid out.

**Review by axic — CHANGES_REQUESTED** [review]
Can you add a test where it tries to pop the stack to access the locals?

File: lib/fizzy/stack.hpp [hunk 6]
```diff
@@ -61,37 +67,63 @@ class OperandStack
...
-    }
-
 public:
     /// Default constructor.
     ///
-    /// Based on @p max_stack_height decides if to use small pre-allocated storage or allocate
-    /// large storage.
-    /// Sets the top item pointer to below the stack bottom.
-    explicit OperandStack(size_t max_stack_height)
+    /// Based on required storage space decides if to use small pre-allocated
```
Comment (axic) [inline lib/fizzy/stack.hpp:83]:
```suggestion
    /// Based on required storage space decides to use small pre-allocated
```

File: lib/fizzy/stack.hpp [hunk 7]
```diff
@@ -61,37 +67,63 @@ class OperandStack
...
+    /// storage or allocate large storage.
+    /// Sets the top stack operand pointer to below the operand stack bottom.
+    /// @param args                 Function arguments. Values are copied at the beginning of the
+    /// storage space.
+    /// @param num_local_variables  The number of the function local variables (excluding
+    ///                             arguments). This number of values is zeroed in the storage space
+    ///                             after the arguments.
+    /// @param max_stack_height     The maximum operand stack height in the function.
```
Comment (axic) [inline lib/fizzy/stack.hpp:91]: Is this the case?
```suggestion
    /// @param max_stack_height     The maximum operand stack height in the function. This excludes args and num_local_variables.
```

File: test/unittests/stack_test.cpp [hunk 8]
```diff
@@ -211,6 +211,13 @@ TEST(operand_stack, small_with_locals)
...
     EXPECT_EQ(stack.local(3).i64, 0xc3);
     EXPECT_EQ(stack.local(4).i64, 0xc4);
+
+    EXPECT_EQ(stack.pop().i64, 0xff);
+    EXPECT_EQ(stack.local(0).i64, 0xc0);
+    EXPECT_EQ(stack.local(1).i64, 0xc1);
+    EXPECT_EQ(stack.local(2).i64, 0xc2);
+    EXPECT_EQ(stack.local(3).i64, 0xc3);
+    EXPECT_EQ(stack.local(4).i64, 0xc4);
 }
```
Comment (axic) [inline test/unittests/stack_test.cpp:221]: Can you also check here locals are unchanged?

File: lib/fizzy/stack.hpp [hunk 9]
```diff
@@ -61,37 +67,63 @@ class OperandStack
     /// in the constructor after the m_storage.
     Value* m_top;
 
+    Value* m_bottom;
+
+    Value* m_locals;
```
Comment (axic) [inline lib/fizzy/stack.hpp:72]: Would the order of these three manifest speed differences?

File: lib/fizzy/stack.hpp [hunk 10]
```diff
@@ -61,37 +67,63 @@ class OperandStack
...
+    /// @param num_local_variables  The number of the function local variables (excluding
+    ///                             arguments). This number of values is zeroed in the storage space
+    ///                             after the arguments.
+    /// @param max_stack_height     The maximum operand stack height in the function. This excludes args and num_local_variables.
+    OperandStack(span<const Value> args, size_t num_local_variables, size_t max_stack_height)
     {
-        if (max_stack_height > small_storage_size)
-            m_large_storage = std::make_unique<Value[]>(max_stack_height);
-        m_top = bottom() - 1;
+        const auto num_args = args.size();
```
Comment (axic) [inline lib/fizzy/stack.hpp:94]: Perhaps this one could make sense to be typed as `size_t` because then we can be sure `storage_size_required` is `size_t`?  Though I guess `size()` returns that anyhow.

File: lib/fizzy/stack.hpp [hunk 11]
```diff
@@ -61,37 +67,63 @@ class OperandStack
...
+        assert(m_locals + index < m_bottom);
+        return m_locals[index];
+    }
+
     /// The current number of items on the stack (aka stack height).
-    size_t size() const noexcept { return static_cast<size_t>(m_top + 1 - bottom()); }
+    size_t size() const noexcept { return static_cast<size_t>(m_top + 1 - m_bottom); }
 
     /// Returns the reference to the top item.
     /// Requires non-empty stack.
```
Comment (axic) [inline lib/fizzy/stack.hpp:129]: Why use `auto` on `top`, `operator[]`, `pop`, when `local`, `rbegin`, and `rend` use the explicit `Value`?

File: lib/fizzy/stack.hpp [hunk 12]
```diff
@@ -49,6 +50,11 @@ class Stack
     }
 };
 
+
+/// Contains current frame's locals (including arguments) and operand stack.
+/// The storage space for locals and operand stack together is allocated as a
+/// continuous memory. Elements occupy the storage space in the order:
+/// arguments, local variables, operand stack.
```
Comment (axic) [inline lib/fizzy/stack.hpp:57]:
```suggestion
/// arguments, local variables, operand stack. Arguments and local variables can be accessed under a single namespace and are separate from the stack itself.
```

File: lib/fizzy/stack.hpp (hunk 10 above)
Comment (chfast) [reply lib/fizzy/stack.hpp:94]: This is fine now, but probably should be changed to `uint32_t` at some point.

File: test/unittests/stack_test.cpp (hunk 8 above)
Comment (chfast) [reply test/unittests/stack_test.cpp:221]: These checks check that locals are unchanged.

File: lib/fizzy/stack.hpp [hunk 13]
```diff
@@ -61,49 +69,76 @@ class OperandStack
...
     ///
-    /// Based on @p max_stack_height decides if to use small pre-allocated storage or allocate
-    /// large storage.
-    /// Sets the top item pointer to below the stack bottom.
-    explicit OperandStack(size_t max_stack_height)
+    /// Based on required storage space decides to use small pre-allocated
+    /// storage or allocate large storage.
+    /// Sets the top stack operand pointer to below the operand stack bottom.
+    /// @param args                 Function arguments. Values are copied at the beginning of the
+    /// storage space.
```
Comment (gumb0) [inline lib/fizzy/stack.hpp:89]: Indentation is off

File: lib/fizzy/stack.hpp (hunk 11 above)
Comment (gumb0) [reply lib/fizzy/stack.hpp:132]: I would prefer explicit `Value` in all cases

Comment (chfast) [reply lib/fizzy/stack.hpp:132]: This is fixed - no more auto.

File: test/unittests/stack_test.cpp [hunk 14]
```diff
@@ -173,10 +181,50 @@ TEST(operand_stack, small)
...
+    stack.local(4) = 0xc4;
+
+    EXPECT_EQ(stack.local(0).i64, 0xc0);
+    EXPECT_EQ(stack.local(1).i64, 0xc1);
+    EXPECT_EQ(stack.local(2).i64, 0xc2);
+    EXPECT_EQ(stack.local(3).i64, 0xc3);
+    EXPECT_EQ(stack.local(4).i64, 0xc4);
+
+    EXPECT_EQ(stack.pop().i64, 0xff);
+    EXPECT_EQ(stack.local(0).i64, 0xc0);
```
Comment (axic) [inline test/unittests/stack_test.cpp:216]: Can you add `size() == 0` here ?

File: lib/fizzy/stack.hpp (hunk 9 above)
Comment (chfast) [reply lib/fizzy/stack.hpp:75]: I have corrected this. New version may be slightly faster, but hard to tell for sure as they are very close.

**Comment by chfast** [comment]
> Can you add a test where it tries to pop the stack to access the locals?

Extended existing tests to have this check.

**Review by gumb0 — APPROVED** [review]

File: test/unittests/stack_test.cpp [hunk 15]
```diff
@@ -136,7 +134,7 @@ TEST(operand_stack, top)
     EXPECT_EQ(stack.top().i64, 101);
     EXPECT_EQ(stack[0].i64, 101);
 
-    stack.shrink(0);
```
Comment (axic) [inline test/unittests/stack_test.cpp:139]: Hm, this reminds me don't have a test for `stack.drop(0);`, it should be a no op. Perhaps just add here before `drop(1)`?

File: test/unittests/stack_test.cpp [hunk 16]
```diff
@@ -117,8 +117,6 @@ TEST(operand_stack, construct)
 {
     OperandStack stack(0);
     EXPECT_EQ(stack.size(), 0);
-    stack.shrink(0);
```
Comment (axic) [inline test/unittests/stack_test.cpp:120]: Hm, this reminds me don't have a test for `stack.drop(0);`, it should be a no op. Perhaps just add here before `drop(1)`?

**Review by axic — APPROVED** [review]
